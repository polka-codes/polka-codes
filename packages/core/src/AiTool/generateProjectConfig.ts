/**
 * AI tool for analyzing project files and generating polkacodes config sections.
 * Generated by polka.codes
 */

import type { AiToolDefinition } from './types'

const prompt = `Analyze the provided project files and generate a polkacodes configuration that captures:

1. Build tools and package manager (e.g., bun, npm)
2. Testing framework and patterns
3. Code style tools and rules
4. Project structure and conventions
5. Common development workflows

Generate a YAML configuration with:
1. scripts section based on package.json scripts and CI workflows
2. rules section based on project conventions, tools, and patterns

The configuration should be enclosed in <tool_output> tags and follow this format:

\`\`\`yaml
scripts:
  test:
    command: "bun test"
    description: "Run tests with bun:test"
  lint:
    command: "biome check ."
    description: "Check code style with Biome"

rules:
  - "Use \`bun\` as package manager"
  - "Write tests using bun:test with snapshots"
  - "Follow Biome code style"
\`\`\`

Focus on accurately reflecting:
- Package manager and dependency management
- Testing frameworks and patterns
- Code style and linting rules
- File organization and naming conventions
- Build and development workflows
`

export default {
  name: 'generateProjectConfig',
  description: 'Analyzes project files to generate polkacodes config sections',
  prompt,
  formatInput: (params: { files: Record<string, string> }) => {
    const fileContent = Object.entries(params.files)
      .map(([path, content]) => `=== ${path} ===\n${content}`)
      .join('\n\n')
    return `Here are the project files to analyze:\n\n${fileContent}`
  },
  parseOutput: (output: string) => {
    const regex = /<tool_output>([\s\S]*)<\/tool_output>/gm
    const match = regex.exec(output)
    if (!match) {
      throw new Error(`Could not parse output:\n${output}`)
    }
    return match[1].trim()
  },
} as const satisfies AiToolDefinition<{ files: Record<string, string> }>
