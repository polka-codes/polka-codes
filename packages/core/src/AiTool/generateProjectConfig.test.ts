/**
 * Tests for generateProjectConfig AiTool
 * Generated by polka.codes
 */

import { describe, expect, test } from 'bun:test'
import generateProjectConfig from './generateProjectConfig'

describe('generateProjectConfig', () => {
  test('should format input correctly', () => {
    const files = {
      'package.json': '{"scripts": {"test": "bun test"}}',
      '.polkacodes.yml': 'rules: ["Use bun"]',
    }

    const input = generateProjectConfig.formatInput({ files })
    expect(input).toMatchSnapshot()
  })

  test('should parse output correctly', () => {
    const output = `<tool_output>
scripts:
  test:
    command: "bun test"
    description: "Run tests"

rules:
  - "Use bun as package manager"
</tool_output>`

    const result = generateProjectConfig.parseOutput(output)
    expect(result).toMatchSnapshot()
  })

  test('should throw error on invalid output', () => {
    const output = 'Invalid output without tool_output tags'
    expect(() => generateProjectConfig.parseOutput(output)).toThrow()
  })
})
