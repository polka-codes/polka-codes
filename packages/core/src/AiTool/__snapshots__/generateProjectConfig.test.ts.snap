// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateProjectConfig should format input correctly 1`] = `
"<tool_input>
=== package.json ===
{"scripts": {"test": "bun test"}}

=== .polkacodes.yml ===
rules: ["Use bun"]
</tool_input>"
`;

exports[`generateProjectConfig should parse output correctly 1`] = `
"scripts:
  test:
    command: "bun test"
    description: "Run tests"

rules:
  - "Use bun as package manager""
`;
